File: internal/handler/redirect/internal_redirect.go
package redirect

import (
	"net/url"
	"regexp"
	"strings"

	"github.com/gin-gonic/gin"
)

// InternalRedirect handles safe redirects within the application
type InternalRedirect struct {
	allowedHosts []string
}

// NewInternalRedirect creates a new instance of InternalRedirect
func NewInternalRedirect(allowedHosts []string) *InternalRedirect {
	return &InternalRedirect{
		allowedHosts: allowedHosts,
	}
}

// SafeRedirectPath checks if a path is safe to redirect to
func (i *InternalRedirect) SafeRedirectPath(path string) string {
	if path == "" {
		return ""
	}

	// Verify that the string starts with a '/' and a known route character
	validPathRegex := regexp.MustCompile(`^/[-\w].*$`)
	if !validPathRegex.MatchString(path) {
		return ""
	}

	// Parse the URI
	uri, err := url.Parse(path)
	if err != nil {
		return ""
	}

	// Return the full path for the URI
	return i.FullPathForURI(uri)
}

// SafeRedirectPathForURL checks if a URL is safe to redirect to
func (i *InternalRedirect) SafeRedirectPathForURL(urlStr string, ctx *gin.Context) string {
	if urlStr == "" {
		return ""
	}

	// Parse the URL
	uri, err := url.Parse(urlStr)
	if err != nil {
		return ""
	}

	// Check if the host is allowed
	if i.HostAllowed(uri, ctx) {
		return i.SafeRedirectPath(i.FullPathForURI(uri))
	}

	return ""
}

// SanitizeRedirect sanitizes a redirect URL or path
func (i *InternalRedirect) SanitizeRedirect(urlOrPath string, ctx *gin.Context) string {
	// Try to sanitize as a path first
	safePath := i.SafeRedirectPath(urlOrPath)
	if safePath != "" {
		return safePath
	}

	// If not a safe path, try to sanitize as a URL
	return i.SafeRedirectPathForURL(urlOrPath, ctx)
}

// HostAllowed checks if a URI's host is allowed
func (i *InternalRedirect) HostAllowed(uri *url.URL, ctx *gin.Context) bool {
	// Get the request host and port
	requestHost := ctx.Request.Host
	hostParts := strings.Split(requestHost, ":")
	requestHostName := hostParts[0]
	requestPort := "80" // Default HTTP port
	if len(hostParts) > 1 {
		requestPort = hostParts[1]
	}

	// Check if the URI host matches the request host
	if uri.Host != requestHostName {
		return false
	}

	// Check if the URI port matches the request port
	uriPort := uri.Port()
	if uriPort == "" {
		// If no port is specified, use the default port based on the scheme
		if uri.Scheme == "https" {
			uriPort = "443"
		} else {
			uriPort = "80"
		}
	}

	return uriPort == requestPort
}

// FullPathForURI returns the full path for a URI
func (i *InternalRedirect) FullPathForURI(uri *url.URL) string {
	// Combine the path and query
	pathWithQuery := uri.Path
	if uri.RawQuery != "" {
		pathWithQuery += "?" + uri.RawQuery
	}

	// Add the fragment if present
	if uri.Fragment != "" {
		pathWithQuery += "#" + uri.Fragment
	}

	return pathWithQuery
}

// RefererPath returns the path from the referer header
func (i *InternalRedirect) RefererPath(ctx *gin.Context) string {
	referer := ctx.Request.Referer()
	if referer == "" {
		return ""
	}

	uri, err := url.Parse(referer)
	if err != nil {
		return ""
	}

	return uri.Path
}

File: rust_app/src/controllers/concerns/internal_redirect.rs
// Ported from: orig_app/app/controllers/concerns/internal_redirect.rb
// Provides safe redirect path utilities for actix-web controllers.
use actix_web::HttpRequest;
use url::Url;

/// Verifies that the path starts with a '/' and a known route character.
pub fn safe_redirect_path(path: &str) -> Option<String> {
    if path.is_empty() {
        return None;
    }
    let re = regex::Regex::new(r"^/[-\w].*").unwrap();
    if !re.is_match(path) {
        return None;
    }
    // Try to parse as a relative URL
    let uri = Url::parse(&format!("http://localhost{}", path)).ok()?;
    Some(full_path_for_uri(&uri))
}

/// Checks if the host and port of the uri match the request's host and port.
pub fn host_allowed(uri: &Url, req: &HttpRequest) -> bool {
    let host = req.connection_info().host();
    let port = uri.port_or_known_default().unwrap_or(80);
    let req_url = format!("http://{}", host);
    if let Ok(req_uri) = Url::parse(&req_url) {
        uri.host_str() == req_uri.host_str()
            && port == req_uri.port_or_known_default().unwrap_or(80)
    } else {
        false
    }
}

/// Returns a safe redirect path for a full URL if the host is allowed.
pub fn safe_redirect_path_for_url(url: &str, req: &HttpRequest) -> Option<String> {
    let uri = Url::parse(url).ok()?;
    if host_allowed(&uri, req) {
        safe_redirect_path(&full_path_for_uri(&uri))
    } else {
        None
    }
}

/// Sanitizes a redirect by checking both path and url forms.
pub fn sanitize_redirect(url_or_path: &str, req: &HttpRequest) -> Option<String> {
    safe_redirect_path(url_or_path).or_else(|| safe_redirect_path_for_url(url_or_path, req))
}

/// Returns the path, query, and fragment from a Url as a single string.
pub fn full_path_for_uri(uri: &Url) -> String {
    let mut path_with_query = uri.path().to_string();
    if let Some(q) = uri.query() {
        path_with_query.push('?');
        path_with_query.push_str(q);
    }
    if let Some(frag) = uri.fragment() {
        path_with_query.push('#');
        path_with_query.push_str(frag);
    }
    path_with_query
}

/// Returns the referer path from the request, if present.
pub fn referer_path(req: &HttpRequest) -> Option<String> {
    req.headers()
        .get("referer")
        .and_then(|v| v.to_str().ok())
        .and_then(|referer| Url::parse(referer).ok())
        .map(|uri| uri.path().to_string())
}
